lib/plugins/iob.js

```diff
@@ -12,22 +12,86 @@ function init() {
     , pluginType: 'pill-major'
   };
 
+  iob.RECENCY_THRESHOLD = 30 * 60 * 1000;
+
   iob.setProperties = function setProperties(sbx) {
     sbx.offerProperty('iob', function setIOB ( ) {
-      return iob.calcTotal(sbx.data.treatments, sbx.data.profile, sbx.time);
+      return iob.calcTotal(sbx.data.treatments, sbx.data.devicestatus, sbx.data.profile, sbx.time);
     });
   };
 
-  iob.calcTotal = function calcTotal(treatments, profile, time, spec_profile) {
+  iob.calcTotal = function calcTotal(treatments, devicestatus, profile, time, spec_profile) {
+    if (time === undefined) {
+      time = Date.now();
+    }
 
-    var totalIOB = 0
-      , totalActivity = 0;
+    var result = iob.lastIOBDeviceStatus(devicestatus, time);
+
+    if (_.isEmpty(result) && treatments !== undefined && treatments.length) {
+      result = iob.fromTreatments(treatments, profile, time, spec_profile);
+    }
 
-    if (!treatments) { return {}; }
+    return addDisplay(result);
+  };
 
-    if (time === undefined) {
-      time = Date.now();
+  function addDisplay(iob) {
+    if (_.isEmpty(iob) || iob.iob === undefined) {
+      return {};
+    }
+    var display = utils.toFixed(iob.iob);
+    return _.merge(iob, {
+      display: display
+      , displayLine: 'IOB: ' + display + 'U'
+    });
+  }
+
+  iob.lastIOBDeviceStatus = function lastIOBDeviceStatus(devicestatus, time) {
+    var recentMills = time - iob.RECENCY_THRESHOLD;
+    return _.chain(devicestatus)
+      .filter(function (devicestatusEntry) {
+        return devicestatusEntry.mills <= time && devicestatusEntry.mills >= recentMills;
+      })
+      .map(iob.fromDeviceStatus)
+      .reject(_.isEmpty)
+      .sortBy('mills')
+      .last( )
+      .value( );
+  };
+
+  iob.fromDeviceStatus = function fromDeviceStatus(devicestatusEntry) {
+    if (_.get(devicestatusEntry, 'openaps.iob.iob') !== undefined) {
+
+      //hacks to support AMA iob array with time fields instead of timestamp fields
+      var iob = _.isArray(devicestatusEntry.openaps.iob) ? devicestatusEntry.openaps.iob[0] : devicestatusEntry.openaps.iob;
+      if (iob.time) {
+        iob.timestamp = iob.time;
+      }
+
+      return {
+        iob: iob.iob
+        , basaliob: iob.basaliob
+        , activity: iob.activity
+        , source: 'OpenAPS'
+        , device: devicestatusEntry.device
+        , mills: moment(iob.timestamp).valueOf( )
+      };
+    } else if (_.get(devicestatusEntry, 'pump.iob') !== undefined) {
+      var iob = devicestatusEntry.pump.iob.iob;
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Variable 'iob' is already declared in line 65.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/2330702123/issues/source?bid=2452379&fileBranchId=3207603#l79)